network_analysis/nbconverted/2_calculate_node_statistics.py

```diff
@@ -0,0 +1,182 @@
+# 
+# * node (gene) degree in the eADAGE graph
+# * edge weight in the eADAGE graph
+# * [betweenness centrality](https://en.wikipedia.org/wiki/Betweenness_centrality) of generic vs. non-generic genes
+# * [PageRank](https://en.wikipedia.org/wiki/PageRank) (sometimes called PageRank centrality) of generic vs. non-generic genes, specifically the [undirected version](https://en.wikipedia.org/wiki/PageRank#PageRank_of_an_undirected_graph).
+
+# In[1]:
+
+
+import os
+
+import numpy as np
+import pandas as pd
+import graph_tool.all as gt
+import matplotlib.pyplot as plt
+import seaborn as sns
+
+gt.seed_rng(1)
+np.random.seed(1)
+
+
+# In[2]:
+
+
+# relevant file paths
+data_dir = './data'
+processed_graph = os.path.join(data_dir, 'eadage_generic_graph_unsigned.gt')
+
+
+# In[3]:
+
+
+G = gt.load_graph(processed_graph)
+# make sure vertex/edge properties exist
+print(G)
+print(list(G.vp.keys()))
+print(list(G.ep.keys()))
+
+
+# In[4]:
+
+
+# plot degree distribution + look at degrees of generic genes
+vs = G.get_vertices()
+names = [G.vp['name'][v] for v in vs]
+is_generic = [G.vp['is_generic'][v] for v in vs]
+degrees = G.get_total_degrees(vs)
+degree_df = pd.DataFrame({
+    'gene': names,
+    'degree': degrees,
+    'is_generic': is_generic
+})
+degree_df.sort_values(by='degree', ascending=False).head()
+
+
+# In[5]:
+
+
+sns.set({'figure.figsize': (12, 4)})
+sns.set_style('whitegrid')
+fig, axarr = plt.subplots(1, 2)
+sns.histplot(data=degree_df, x='degree', hue='is_generic', element='step',
+             stat='probability', common_norm=False, ax=axarr[0])
+axarr[0].set_title('Degree distribution of generic/non-generic genes')
+sns.boxplot(data=degree_df, y='degree', x='is_generic', ax=axarr[1])
+axarr[1].set_title('Degree distribution of generic/non-generic genes')
+
+
+# In[6]:
+
+
+# plot weight distribution + look at weights of edges including generic genes
+# (v1, v2, weight, is_generic)
+edges = []
+for s, t, w in G.iter_edges([G.ep['weight']]):
+    if (G.vp['is_generic'][s]) and (G.vp['is_generic'][t]):
+        is_generic = 2
+    elif (G.vp['is_generic'][s]) or (G.vp['is_generic'][t]):
+        is_generic = 1
+    else:
+        is_generic = 0
+    edges.append((G.vp['name'][s], G.vp['name'][t], w, is_generic))
+weight_df = pd.DataFrame(edges, columns=['g1', 'g2', 'weight', 'is_generic'])
+print(weight_df.shape)
+weight_df.head()
+
+
+# In[7]:
+
+
+sns.set({'figure.figsize': (12, 4)})
+sns.set_style('whitegrid')
+fig, axarr = plt.subplots(1, 2)
+sns.histplot(data=weight_df, x='weight', hue='is_generic', element='step',
+             stat='probability', common_norm=False, ax=axarr[0])
+axarr[0].set_title('Weight distribution of generic/non-generic edges')
+sns.boxplot(data=weight_df, y='weight', x='is_generic', ax=axarr[1])
+axarr[1].set_title('Weight distribution of generic/non-generic edges')
```

Comment:
So i think that bump means that there are some generic genes that are more likely to be high weight with other generic genes than any other gene..

Reply:
Yeah, Casey and I were talking about this in our meeting yesterday - I agree with your interpretation. We think it could be interesting to see if that bump disappears if we use a coexpression network for these analyses (i.e. if generic genes are only very similar when the similarity is calculated using eADAGE LVs, as opposed to just using gene expression over all genes).

Not sure if we want to go further with this analysis, though, or if what we've done is enough for the current paper - I think it's up to you and Casey. Happy to chat about this over Slack if that would be better.